### PR TITLE
element index init with vptr

### PR DESCRIPTION
**Pull request by dwblaikie**
Not sure if this is the most robust way to do it - I guess the alternative is doing name lookup into the dest struct fields too?

### CODE REVIEW

**Comment by dwblaikie** [comment]
This closes the other gap mentioned in https://github.com/carbon-language/carbon-lang/pull/4514 (#4515 addressed this issue in a way that #4515 did not/left unanswered)

File: toolchain/check/convert.cpp [hunk 1]
```diff
@@ -183,15 +183,16 @@ static auto ConvertAggregateElement(
     SemIR::TypeId src_elem_type,
     llvm::ArrayRef<SemIR::InstId> src_literal_elems,
     ConversionTarget::Kind kind, SemIR::InstId target_id,
-    SemIR::TypeId target_elem_type, PendingBlock* target_block, size_t i) {
+    SemIR::TypeId target_elem_type, PendingBlock* target_block,
+    size_t src_field_index, size_t dest_field_index) {
```
Comment (geoffromer) [inline toolchain/check/convert.cpp]: Looks like the comment needs to be updated to reflect these changes.

File: toolchain/check/convert.cpp [hunk 2]
```diff
@@ -496,7 +498,7 @@ static auto ConvertStructToStructOrClass(Context& context,
         ConvertAggregateElement<SemIR::StructAccess, TargetAccessInstT>(
             context, value_loc_id, value_id, src_field.type_id, literal_elems,
             inner_kind, target.init_id, dest_field.type_id, target.init_block,
-            src_field_index);
+            src_field_index, src_field_index + dest_has_vptr);
```
Comment (geoffromer) [inline toolchain/check/convert.cpp]:
```suggestion
            src_field_index, src_field_index + dest_has_vptr ? 1 : 0);
```

Whenever I see code that treats a boolean as a number, I have to stop and think through what the rules for that are.

File: toolchain/check/testdata/class/virtual_modifiers.carbon [hunk 3]
```diff
@@ -235,3 +249,105 @@ fn F() {
...
+// CHECK:STDOUT:   %F.decl: %F.type.2 = fn_decl @F.2 [template = constants.%F.2] {} {}
+// CHECK:STDOUT: }
+// CHECK:STDOUT:
+// CHECK:STDOUT: class @Base {
+// CHECK:STDOUT:   %F.decl: %F.type.1 = fn_decl @F.1 [template = constants.%F.1] {} {}
+// CHECK:STDOUT:   %.loc6_11.1: Core.IntLiteral = int_value 32 [template = constants.%.1]
+// CHECK:STDOUT:   %int.make_type_signed.loc6: init type = call constants.%Int(%.loc6_11.1) [template = constants.%i32]
+// CHECK:STDOUT:   %.loc6_11.2: type = value_of_initializer %int.make_type_signed.loc6 [template = constants.%i32]
+// CHECK:STDOUT:   %.loc6_11.3: type = converted %int.make_type_signed.loc6, %.loc6_11.2 [template = constants.%i32]
+// CHECK:STDOUT:   %.loc6_9: %.2 = field_decl m1, element0 [template]
```
Comment (geoffromer) [inline toolchain/check/testdata/class/virtual_modifiers.carbon:309]: It looks like the vptr isn't included in the indexing scheme for `field_decl`, but it is for `class_element_access`. Is that intended?

Comment (dwblaikie) [reply toolchain/check/testdata/class/virtual_modifiers.carbon:309]: Yeah, that should be addressed by #4514

File: toolchain/check/convert.cpp (hunk 1 above)
Comment (dwblaikie) [reply toolchain/check/convert.cpp]: Ah, yep - done in https://github.com/carbon-language/carbon-lang/pull/4565/commits/fafd925c8a353c1b0a829845c86f3920d8071b1e

File: toolchain/check/convert.cpp (hunk 2 above)
Comment (dwblaikie) [reply toolchain/check/convert.cpp]: Ah, sure - there was another use of this already, so I addressed both by introducing a `dest_vptr_offset` integer local to use in both places. b68ab0d6cb171bd30af67d143424169353ec327c

**Review by geoffromer — APPROVED** [review]

**Comment by geoffromer** [comment]
Looks like the tests are failing

**Pull request merged**